### PR TITLE
Generate markdown help docs when generating template schemas.

### PR DESCRIPTION
**Pull request by jhendrixMSFT**
The AzureResourceSchema generator will now generate a help document
alongside the generated resource schema.  The doc contains an example
schema and tables describing the various properties; this is modeled from
the existing help docs that are hand-written.

### CODE REVIEW

**Comment by jhendrixMSFT** [comment]
@tbombach Here's the PR for the markdown generator.

**Comment by tbombach** [comment]
Awesome, look good to me! @daschult any concerns before I merge?

File: src/core/AutoRest.Core/CodeGenerator.cs [hunk 1]
```diff
@@ -139,7 +140,7 @@ public virtual void PopulateSettings(IDictionary<string, object> settings)
                     {
                         await textWriter.WriteLineAsync();
                     }
-                    else if (!string.IsNullOrWhiteSpace(line))
+                    else if (!skipEmptyLines || (skipEmptyLines && !string.IsNullOrWhiteSpace(line)))
```
Comment (daschult) [inline src/core/AutoRest.Core/CodeGenerator.cs]: The second check for skipEmptyLines is unnecessary since it can only be true if the second expression is evaluated.

File: src/generator/AutoRest.AzureResourceSchema/AzureResourceSchemaCodeGenerator.cs [hunk 2]
```diff
@@ -50,7 +50,13 @@ public override void NormalizeClientModel(ServiceClient serviceClient)
                 ResourceSchemaWriter.Write(stringWriter, resourceSchemas[resourceProvider]);
 
                 string schemaPath = Path.Combine(Settings.OutputDirectory, resourceProvider + ".json");
-                await Write(stringWriter.ToString(), schemaPath);
+                await Write(stringWriter.ToString(), schemaPath, true);
+
+                stringWriter = new StringWriter();
+                ResourceMarkdownWriter.Write(stringWriter, resourceSchemas[resourceProvider]);
```
Comment (daschult) [inline src/generator/AutoRest.AzureResourceSchema/AzureResourceSchemaCodeGenerator.cs]: Why did you choose to make this part of the AzureResourceSchema generator, and not its own generator?

File: src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs [hunk 3]
```diff
@@ -13,6 +13,7 @@ namespace AutoRest.AzureResourceSchema
     /// </summary>
     public class JsonSchema
     {
+        private string resourceType;
```
Comment (daschult) [inline src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs]: Did you pull this out of the properties dictionary for performance reasons?

File: src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs [hunk 4]
```diff
@@ -48,38 +49,29 @@ public class JsonSchema
...
-                }
-
-                return result;
+                return resourceType;
             }
             set
             {
-                if (properties == null)
+                resourceType = value;
+                if (Properties != null && Properties.ContainsKey("type"))
```
Comment (daschult) [inline src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs]: Previously properties could be initialized in this (ResourceType) property's setter. Now it can't. Is that a regression?

File: src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs [hunk 5]
```diff
@@ -82,6 +82,7 @@ public static class ResourceSchemaParser
                     JsonSchema resourceDefinition = new JsonSchema();
                     resourceDefinition.JsonType = "object";
 
+                    resourceDefinition.ResourceType = resourceType;
```
Comment (daschult) [inline src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs]: Why not just look for the "type" property instead of creating a ResourceType property?

**Comment by fearthecowboy** [comment]
The `RefactorCodeModel ` branch is about to be merged; this will have to wait until that branch is merged to `master`.

File: src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs (hunk 5 above)
Comment (jhendrixMSFT) [inline src/generator/AutoRest.AzureResourceSchema/ResourceSchemaParser.cs]: This heuristic isn't reliable.  For example there are some swaggers that contain a property named "type" that is unrelated to resource type (I assume this is allowed).

File: src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs (hunk 4 above)
Comment (jhendrixMSFT) [inline src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs]: No, we still initialize it as required in the AddProperty() method.

File: src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs (hunk 3 above)
Comment (jhendrixMSFT) [inline src/generator/AutoRest.AzureResourceSchema/JsonSchema.cs]: No, I removed it because relying on an entry with the name "type" is ambiguous (see my other comment regarding setting this value).

**Comment by jhendrixMSFT** [comment]
@fearthecowboy Is the merge complete?

**Comment by fearthecowboy** [comment]
If this passes tests and merges cleanly against master, I won't block it now :D

**Comment by jhendrixMSFT** [comment]
@daschult Do you have any more comments?

**Comment by daschult** [comment]
No, go ahead and merge it.

**Comment by jhendrixMSFT** [comment]
@fearthecowboy If you're good with the changes could you please merge?
